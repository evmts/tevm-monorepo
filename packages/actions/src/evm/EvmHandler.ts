import type { Quantity } from 'viem'

/**
 * EVM handler for `evm_setNextBlockTimestamp` method
 */
export type EvmSetNextBlockTimestampHandler = (
	params: readonly [timestamp: Quantity],
) => Promise<{ errors?: never } | { errors: Error[] }>

/**
 * Mapping of `evm_*` method names to their respective handlers
 */
export type EvmHandler = {
	evm_setNextBlockTimestamp: EvmSetNextBlockTimestampHandler
}