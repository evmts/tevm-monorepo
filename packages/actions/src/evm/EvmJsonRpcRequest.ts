import type { Quantity } from 'viem'
import type { JsonRpcRequest } from 'viem'

/**
 * JSON-RPC request for `evm_setNextBlockTimestamp` method
 */
export type EvmSetNextBlockTimestampJsonRpcRequest = JsonRpcRequest<
	'evm_setNextBlockTimestamp',
	readonly [timestamp: Quantity]
>

/**
 * Union of all EVM JSON-RPC requests
 */
export type EvmJsonRpcRequest = EvmSetNextBlockTimestampJsonRpcRequest