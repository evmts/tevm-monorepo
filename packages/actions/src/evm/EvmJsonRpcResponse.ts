import type { JsonRpcResponse } from 'viem'

/**
 * JSON-RPC response for `evm_setNextBlockTimestamp` method
 */
export type EvmSetNextBlockTimestampJsonRpcResponse = JsonRpcResponse<null, void, 'evm_setNextBlockTimestamp'>

/**
 * Union of all EVM JSON-RPC responses
 */
export type EvmJsonRpcResponse = EvmSetNextBlockTimestampJsonRpcResponse