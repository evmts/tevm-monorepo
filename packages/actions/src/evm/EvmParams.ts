import type { Quantity } from 'viem'

/**
 * Parameters for `evm_setNextBlockTimestamp` method
 */
export type EvmSetNextBlockTimestampParams = readonly [timestamp: Quantity]

/**
 * Union of all EVM method parameters
 */
export type EvmParams = {
	evm_setNextBlockTimestamp: EvmSetNextBlockTimestampParams
}