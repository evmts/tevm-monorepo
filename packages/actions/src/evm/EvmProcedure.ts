import type { EvmSetNextBlockTimestampJsonRpcRequest, EvmSetNextBlockTimestampJsonRpcResponse } from '.'

/**
 * Procedure type for `evm_setNextBlockTimestamp` method
 */
export type EvmSetNextBlockTimestampProcedure = (
	request: EvmSetNextBlockTimestampJsonRpcRequest,
) => Promise<EvmSetNextBlockTimestampJsonRpcResponse>

/**
 * Union of all EVM procedure types
 */
export type EvmProcedure = {
	evm_setNextBlockTimestamp: EvmSetNextBlockTimestampProcedure
}