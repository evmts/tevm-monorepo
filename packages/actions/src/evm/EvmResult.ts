/**
 * Result for `evm_setNextBlockTimestamp` method
 */
export type EvmSetNextBlockTimestampResult = void

/**
 * Union of all EVM method results
 */
export type EvmResult = {
	evm_setNextBlockTimestamp: EvmSetNextBlockTimestampResult
}