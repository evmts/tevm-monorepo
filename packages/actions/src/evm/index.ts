export type {
	EvmHandler,
	EvmSetNextBlockTimestampHandler,
} from './EvmHandler.js'
export type {
	EvmJsonRpcRequest,
	EvmSetNextBlockTimestampJsonRpcRequest,
} from './EvmJsonRpcRequest.js'
export type {
	EvmJsonRpcResponse,
	EvmSetNextBlockTimestampJsonRpcResponse,
} from './EvmJsonRpcResponse.js'
export type {
	EvmParams,
	EvmSetNextBlockTimestampParams,
} from './EvmParams.js'
export type {
	EvmProcedure,
	EvmSetNextBlockTimestampProcedure,
} from './EvmProcedure.js'
export type {
	EvmResult,
	EvmSetNextBlockTimestampResult,
} from './EvmResult.js'

export { evmSetNextBlockTimestampProcedure } from './evmSetNextBlockTimestampProcedure.js'