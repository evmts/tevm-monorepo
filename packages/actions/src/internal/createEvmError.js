import {
	AuthCallUnsetError,
	BLS12381FpNotInFieldError,
	BLS12381InputEmptyError,
	BLS12381PointNotOnCurveError,
	BaseError,
	CodeSizeExceedsMaximumError,
	CodeStoreOutOfGasError,
	CreateCollisionError,
	InitcodeSizeViolationError,
	InsufficientBalanceError,
	InternalError,
	InvalidBeginSubError,
	InvalidBytecodeResultError,
	InvalidCommitmentError,
	InvalidEofFormatError,
	InvalidInputLengthError,
	InvalidJumpError,
	InvalidJumpSubError,
	InvalidKzgInputsError,
	InvalidOpcodeError,
	InvalidProofError,
	InvalidReturnSubError,
	OutOfGasError,
	OutOfRangeError,
	RefundExhaustedError,
	RevertError,
	StackOverflowError,
	StackUnderflowError,
	StaticStateChangeError,
	StopError,
	ValueOverflowError,
} from '@tevm/errors'
import { EvmErrorMessage } from '@tevm/evm'

/**
 * @type {ReturnType<typeof createEvmError>} EvmError
 */

/**
 * @param {import('@tevm/evm').EvmError} error
 */
export const createEvmError = (error) => {
	// make it so we can always override it for now
	if (error instanceof BaseError) {
		return /** @type {never}*/ (error)
	}
	switch (error.error) {
		case EvmErrorMessage.STOP: {
			return new StopError(error.error, { cause: error })
		}
		case EvmErrorMessage.REVERT: {
			return new RevertError(error.error, { cause: error })
		}
		case EvmErrorMessage.OUT_OF_GAS: {
			return new OutOfGasError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_OPCODE: {
			return new InvalidOpcodeError(error.error, { cause: error })
		}
		case EvmErrorMessage.STACK_OVERFLOW: {
			return new StackOverflowError(error.error, { cause: error })
		}
		case EvmErrorMessage.STACK_UNDERFLOW: {
			return new StackUnderflowError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_JUMP: {
			return new InvalidJumpError(error.error, { cause: error })
		}
		case EvmErrorMessage.OUT_OF_RANGE: {
			return new OutOfRangeError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_PROOF: {
			return new InvalidProofError(error.error, { cause: error })
		}
		case EvmErrorMessage.AUTHCALL_UNSET: {
			return new AuthCallUnsetError(error.error, { cause: error })
		}
		case EvmErrorMessage.INTERNAL_ERROR: {
			return new InternalError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_INPUTS: {
			return new InvalidKzgInputsError(error.error, { cause: error })
		}
		case EvmErrorMessage.VALUE_OVERFLOW: {
			return new ValueOverflowError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_JUMPSUB: {
			return new InvalidJumpSubError(error.error, { cause: error })
		}
		case EvmErrorMessage.CREATE_COLLISION: {
			return new CreateCollisionError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_BEGINSUB: {
			return new InvalidBeginSubError(error.error, { cause: error })
		}
		case EvmErrorMessage.REFUND_EXHAUSTED: {
			return new RefundExhaustedError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_RETURNSUB: {
			return new InvalidReturnSubError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_COMMITMENT: {
			return new InvalidCommitmentError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_EOF_FORMAT: {
			return new InvalidEofFormatError(error.error, { cause: error })
		}
		case EvmErrorMessage.STATIC_STATE_CHANGE: {
			return new StaticStateChangeError(error.error, { cause: error })
		}
		case EvmErrorMessage.CODESTORE_OUT_OF_GAS: {
			return new CodeStoreOutOfGasError(error.error, { cause: error })
		}
		case EvmErrorMessage.INSUFFICIENT_BALANCE: {
			return new InsufficientBalanceError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_INPUT_LENGTH: {
			return new InvalidInputLengthError(error.error, { cause: error })
		}
		case EvmErrorMessage.BLS_12_381_INPUT_EMPTY: {
			return new BLS12381InputEmptyError(error.error, { cause: error })
		}
		case EvmErrorMessage.INITCODE_SIZE_VIOLATION: {
			return new InitcodeSizeViolationError(error.error, { cause: error })
		}
		case EvmErrorMessage.INVALID_BYTECODE_RESULT: {
			return new InvalidBytecodeResultError(error.error, { cause: error })
		}
		case EvmErrorMessage.CODESIZE_EXCEEDS_MAXIMUM: {
			return new CodeSizeExceedsMaximumError(error.error, { cause: error })
		}
		case EvmErrorMessage.BLS_12_381_FP_NOT_IN_FIELD: {
			return new BLS12381FpNotInFieldError(error.error, { cause: error })
		}
		case EvmErrorMessage.BLS_12_381_POINT_NOT_ON_CURVE: {
			return new BLS12381PointNotOnCurveError(error.error, { cause: error })
		}
		default: {
			return new InternalError(error.error, { cause: error })
		}
	}
}
