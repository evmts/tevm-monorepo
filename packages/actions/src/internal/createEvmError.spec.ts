import {
	AuthCallUnsetError,
	BLS12381FpNotInFieldError,
	BLS12381InputEmptyError,
	BLS12381PointNotOnCurveError,
	CodeSizeExceedsMaximumError,
	CodeStoreOutOfGasError,
	CreateCollisionError,
	InitcodeSizeViolationError,
	InsufficientBalanceError,
	InternalError,
	InvalidBeginSubError,
	InvalidBytecodeResultError,
	InvalidCommitmentError,
	InvalidEofFormatError,
	InvalidInputLengthError,
	InvalidJumpError,
	InvalidJumpSubError,
	InvalidKzgInputsError,
	InvalidOpcodeError,
	InvalidProofError,
	InvalidReturnSubError,
	OutOfGasError,
	OutOfRangeError,
	RefundExhaustedError,
	RevertError,
	StackOverflowError,
	StackUnderflowError,
	StaticStateChangeError,
	StopError,
	ValueOverflowError,
} from '@tevm/errors'
import { EvmErrorMessage } from '@tevm/evm'
import { describe, expect, it } from 'vitest'
import { createEvmError } from './createEvmError.js'

describe('createEvmError', () => {
	const testCases = [
		{ error: { error: EvmErrorMessage.STOP }, expected: StopError },
		{ error: { error: EvmErrorMessage.REVERT }, expected: RevertError },
		{ error: { error: EvmErrorMessage.OUT_OF_GAS }, expected: OutOfGasError },
		{ error: { error: EvmErrorMessage.INVALID_OPCODE }, expected: InvalidOpcodeError },
		{ error: { error: EvmErrorMessage.STACK_OVERFLOW }, expected: StackOverflowError },
		{ error: { error: EvmErrorMessage.STACK_UNDERFLOW }, expected: StackUnderflowError },
		{ error: { error: EvmErrorMessage.INVALID_JUMP }, expected: InvalidJumpError },
		{ error: { error: EvmErrorMessage.OUT_OF_RANGE }, expected: OutOfRangeError },
		{ error: { error: EvmErrorMessage.INVALID_PROOF }, expected: InvalidProofError },
		{ error: { error: EvmErrorMessage.AUTHCALL_UNSET }, expected: AuthCallUnsetError },
		{ error: { error: EvmErrorMessage.INTERNAL_ERROR }, expected: InternalError },
		{ error: { error: EvmErrorMessage.INVALID_INPUTS }, expected: InvalidKzgInputsError },
		{ error: { error: EvmErrorMessage.VALUE_OVERFLOW }, expected: ValueOverflowError },
		{ error: { error: EvmErrorMessage.INVALID_JUMPSUB }, expected: InvalidJumpSubError },
		{ error: { error: EvmErrorMessage.CREATE_COLLISION }, expected: CreateCollisionError },
		{ error: { error: EvmErrorMessage.INVALID_BEGINSUB }, expected: InvalidBeginSubError },
		{ error: { error: EvmErrorMessage.REFUND_EXHAUSTED }, expected: RefundExhaustedError },
		{ error: { error: EvmErrorMessage.INVALID_RETURNSUB }, expected: InvalidReturnSubError },
		{ error: { error: EvmErrorMessage.INVALID_COMMITMENT }, expected: InvalidCommitmentError },
		{ error: { error: EvmErrorMessage.INVALID_EOF_FORMAT }, expected: InvalidEofFormatError },
		{ error: { error: EvmErrorMessage.STATIC_STATE_CHANGE }, expected: StaticStateChangeError },
		{ error: { error: EvmErrorMessage.CODESTORE_OUT_OF_GAS }, expected: CodeStoreOutOfGasError },
		{ error: { error: EvmErrorMessage.INSUFFICIENT_BALANCE }, expected: InsufficientBalanceError },
		{ error: { error: EvmErrorMessage.INVALID_INPUT_LENGTH }, expected: InvalidInputLengthError },
		{ error: { error: EvmErrorMessage.BLS_12_381_INPUT_EMPTY }, expected: BLS12381InputEmptyError },
		{ error: { error: EvmErrorMessage.INITCODE_SIZE_VIOLATION }, expected: InitcodeSizeViolationError },
		{ error: { error: EvmErrorMessage.INVALID_BYTECODE_RESULT }, expected: InvalidBytecodeResultError },
		{ error: { error: EvmErrorMessage.CODESIZE_EXCEEDS_MAXIMUM }, expected: CodeSizeExceedsMaximumError },
		{ error: { error: EvmErrorMessage.BLS_12_381_FP_NOT_IN_FIELD }, expected: BLS12381FpNotInFieldError },
		{ error: { error: EvmErrorMessage.BLS_12_381_POINT_NOT_ON_CURVE }, expected: BLS12381PointNotOnCurveError },
	]

	testCases.forEach(({ error, expected }) => {
		it(`should return ${expected.name} for ${error.error}`, () => {
			const result = createEvmError(error as any)
			expect(result).toBeInstanceOf(expected)
			expect(result.message.includes(error.error)).toBe(true)
		})
	})

	it('should return InternalError for unknown error', () => {
		const error = { error: 'UNKNOWN_ERROR' }
		const result = createEvmError(error as any)
		expect(result).toBeInstanceOf(InternalError)
		expect(result.name).toBe(InternalError.name)
		expect(result.message).toMatchSnapshot()
	})

	it('should return the same error if it is an instance of BaseError', () => {
		const baseError = new InternalError('Existing BaseError')
		const result = createEvmError(baseError as any)
		expect(result).toBe(baseError)
	})
})
