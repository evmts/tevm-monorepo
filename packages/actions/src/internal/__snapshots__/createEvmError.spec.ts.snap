// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createEvmError > should return InternalError for unknown error 1`] = `
[InternalError: UNKNOWN_ERROR

Docs: https://tevm.sh/reference/tevm/errors/classes/internalerror/
Details: {"error":"UNKNOWN_ERROR"}
Version: 1.1.0.next-73]
`;

exports[`createEvmError should return InternalError for unknown error 1`] = `[_InternalError: UNKNOWN_ERROR

Docs: https://tevm.sh/reference/tevm/errors/classes/internalerror/
Details: {"error":"UNKNOWN_ERROR"}
Version: 1.1.0.next-73]`;
